Framework/Algorithms/src/FitPeaks.cpp

```diff
@@ -870,74 +930,133 @@ std::vector<std::shared_ptr<FitPeaksAlgorithm::PeakFitResult>> FitPeaks::fitPeak
+                        size_t &right_index);
+
 /// Supported peak profiles for observation
 std::vector<std::string> supported_peak_profiles{"Gaussian", "Lorentzian", "PseudoVoigt", "Voigt",
                                                  "BackToBackExponential"};
-
+//----------------------------------------------------------------------------------------------
+/** Sum up all counts in a histogram
+ * @param histogram :: histogram instance
+ * @return :: total number of counts
+ */
 double numberCounts(const Histogram &histogram) {
-  double total = 0.;
-  for (const auto &value : histogram.y())
-    total += std::fabs(value);
+  double total = std::accumulate(histogram.y().begin(), histogram.y().end(), 0.);
   return total;
 }
 
 //----------------------------------------------------------------------------------------------
-/** Get number of counts in a specified range of a histogram
+/** Sum up all counts in a specified range of a histogram
  * @param histogram :: histogram instance
- * @param xmin :: left boundary
- * @param xmax :: right boundary
- * @return :: counts
+ * @param range :: histogram range
+ * @return :: total number of counts
  */
-double numberCounts(const Histogram &histogram, const double xmin, const double xmax) {
-  const auto &vector_x = histogram.points();
-
-  // determine left boundary
-  auto start_iter = vector_x.begin();
-  if (xmin > vector_x.front())
-    start_iter = std::lower_bound(vector_x.begin(), vector_x.end(), xmin);
-  if (start_iter == vector_x.end())
-    return 0.; // past the end of the data means nothing to integrate
-  // determine right boundary
-  auto stop_iter = vector_x.end();
-  if (xmax < vector_x.back()) // will set at end of vector if too large
-    stop_iter = std::lower_bound(start_iter, stop_iter, xmax);
-
-  // convert to indexes to sum over y
-  size_t start_index = static_cast<size_t>(start_iter - vector_x.begin());
-  size_t stop_index = static_cast<size_t>(stop_iter - vector_x.begin());
-
-  // integrate
-  double total = 0.;
-  for (size_t i = start_index; i < stop_index; ++i)
-    total += std::fabs(histogram.y()[i]);
+double numberCounts(const Histogram &histogram, const std::pair<double, double> &range) {
+  // get data range
+  std::vector<double> vec_x, vec_y, vec_e;
+  getRangeData(histogram, range, vec_x, vec_y, vec_e);
+  // sum up all counts
+  double total = std::accumulate(vec_y.begin(), vec_y.end(), 0.);
   return total;
 }
 
 //----------------------------------------------------------------------------------------------
-/** Get number of vector elements falling into specified range
- * @param vector :: vector of type T
+/** Estimate background noise given a peak-window Y-vector
+ * @param vec_y :: a peak window of Y-values
+ * @return :: noise estimate
+ */
+double estimateBackgroundNoise(const std::vector<double> &vec_y) {
+  // assume that the first and the last 10 data points are background. Combine them into a data vector.
+  assert(vec_y.size() > 20);
```

Comment:
I don't think  that the fitting window is guaranteed to have at least 20 bins in it, and also it is probably not a good idea to hard-code in a number of bins because the diffractometer resolution $\Delta d/d$ is roughly constant at a given scattering angle (so atr ISIS we determine the peak window width to scale the same way) it might be better to use that first/last 10% of point or 2 bins which ever is greater?

Reply:
This assertion is for an earlier check of how many data points we should have in a peak window.  That check was added recently:

https://github.com/mantidproject/mantid/blob/ED_EWM_245_SNR_peak_checks/Framework/Algorithms/src/FitPeaks.cpp#L1677-L1688

One consideration is: we fit 5 parameters (3 Gaussian and 2 for linear background.)  The general rule of thumb would be to have at least 10 data points per parameter.  Another consideration is: we have an existing `estimateBackgroundParameters` method which implicitly requires to have at least 20 data points: 

https://github.com/mantidproject/mantid/blob/ED_EWM_245_SNR_peak_checks/Framework/Algorithms/src/FitPeaks.cpp#L1585-L1587

I have a question. For the type of experiments you are dealing with, would it be possible to use a smaller step in the binning, so that you have more bins per a peak window?

Reply:
> One consideration is: we fit 5 parameters (3 Gaussian and 2 for linear background.) The general rule of thumb would be to have at least 10 data points per parameter.

Agreed that you need more points than fitting parameters in the window (hopefully many more!) - indeed there are peak functions with more parameters (e.g. `BackToBackExponential`), so there should be a separate check for this (I thought there already was actually) which gets the number of parameters from the function chosen. If the window is too small we could skip the peak or increase the window (probably the former).

> Another consideration is: we have an existing estimateBackgroundParameters method which implicitly requires to have at least 20 data points:

It seems to me we could estimate the background parameters from fewer points and therefore throw away fewer peaks (this might be important for low d-spacing peaks which are narrow and need small windows to avoid overlap with other peak windows). 

`I have a question. For the type of experiments you are dealing with, would it be possible to use a smaller step in the binning, so that you have more bins per a peak window?`

Good question ,unfortunately this is not really an option for two reasons:

1. The minimum bin width is determined by the instrument electronics - we operate in histogram mode so data are already binned in TOF.
2. When fitting an individual detector spectrum we struggle for flux, it is then necessary to increase the bin width (typically double it)  to improve the stats for the fitting so minimizing chi-squared works better (although technically we are still be in the Poisson regime with non-normal errors) - even then we frequently have many empty bins.